### PR TITLE
Backport from go-ipfs, fix tests, hopefully fix Windows

### PR DESCRIPTION
**Pull request by tv42**
Untested on Windows, hoping some kind soul on IRC will do that.

To test, run whatever is the Windows equivalent of:

```
go get github.com/jbenet/go-datastore && cd $GOPATH/src/github.com/jbenet/go-datastore && git remote add tv42 https://github.com/tv42/go-datastore && git fetch tv42 && git checkout flatfs-rides-again && go test ./flatfs
```

### CODE REVIEW

File: flatfs/sync_windows.go [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package flatfs
+
+func syncDir(dir string) error {
+	return nil
```
Comment (jbenet) [inline flatfs/sync_windows.go:4]: so i guess windows doesnt like sync?

**Comment by jbenet** [comment]
@tv42 this all LGTM. Ready For Merge?

**Comment by tv42** [comment]
Now this should be good to merge. Not tested yet on Windows, as far as I know.

**Comment by tv42** [comment]
@gatesvp your help would be greatly appreciated; `go test ./flatfs` again, on Windows.

**Comment by gatesvp** [comment]
Will test this shortly, I am in transit right now. Will double check in a
few hours.

**Comment by gatesvp** [comment]
we are improving :)

`--- FAIL: TestStorage (0.01s)
        flatfs_test.go:186: file should not be world accessible: 0666
FAIL
FAIL    github.com/jbenet/go-datastore/flatfs   0.121s`

**Comment by tv42** [comment]
I guess I should just disable file mode checking on Windows.

**Comment by tv42** [comment]
@gatesvp Like a game of table tennis.

**Comment by gatesvp** [comment]
`ok      github.com/jbenet/go-datastore/flatfs   0.096s`

pong :)

**Comment by tv42** [comment]
Whee! Rebased to make it sit on top of master.

@jbenet, this is ready for merge; once that's good, I'll vendor it into ipfs.

**Comment by jbenet** [comment]
thanks @tv42 and @gatesvp -- merging.

**Pull request merged**